pkg/models/object.go

```diff
@@ -0,0 +1,62 @@
+package models
+
+// RawObject maps pretty closely to raw blob storage
+type RawObject struct {
+	UID  string `json:"uid"`
+	Kind string `json:"kind"`
+
+	// Who saved it when
+	Created   int64 `json:"created,omitempty"`   // timestamp
+	CreatedBy int64 `json:"createdBy,omitempty"` // user_id
+	Updated   int64 `json:"updated,omitempty"`   // timestamp  Last-Modified
+	UpdatedBy int64 `json:"updatedBy,omitempty"` // user_id
+
+	// Raw body
+	Size int64  `json:"size"`           // Content-Length
+	ETag string `json:"etag,omitempty"` // MD5 digest of the Body
+	Body []byte `json:"-"`              // don't return this as JSON unless explicit
+
+	// Optional additional properties managed by the underlying storage
+	Properties map[string]string `json:"properties,omitempty"`
+
+	// Not in every object store, but might be
+	Version     string `json:"version,omitempty"`  // commit hash, incrementing number (as string)
+	SyncTime    int64  `json:"syncTime,omitempty"` // when was the object synced from external source (provisioning or git)
+	SaveMessage string `json:"message,omitempty"`  // the commit message when this was saved
+}
+
+// ObjectSummary is derived from a RawObject and should not depend on system state
+type ObjectSummary struct {
+	Name        string            `json:"name,omitempty"`
+	Description string            `json:"description,omitempty"`
+	Labels      map[string]string `json:"labels,omitempty"` // "tags" are represented as keys with empty values
+	URL         string            `json:"URL,omitempty"`    // not great to save here, but maybe not terrible :shrug:
+	Error       string            `json:"error,omitempty"`  // Reason the object is invalid
+
+	// Optional values -- naming convention for types?
+	Fields map[string]interface{} `json:"fields,omitempty"` // Saved as JSON, returned in results, but values not sortable
+
+	// eg: panels within dashboard
+	Nested []NestedObjectSummary `json:"nested,omitempty"`
```

Comment:
Is this case too extraordinary?

Reply:
is the main use case to search panel titles?

Reply:
The immediate use case is that dashboards actually define two things we want to search for... dashboards and panels within dashboards.  "Title" is part of it, but each panel has all the same reference objects (ie datasource/plugin/etc) so i *think* the nested structure makes sense

Reply:
References are already common, and as we decompose dashboards, we should only expect them to become more common.

I can say that looking at this, and other related types (`NestedObjectSummary`, `ExternalReference`), feels like the product of a bunch of assumptions about how references work.

I'm pretty eager to have those assumptions clearly spelled out, as references bear significantly on the compositions questions that as-code is concerned with. Basically, i'm just going to look at this and feel anxious until i'm able to connect what's here with the abstractions we aim to provide in an as-code context for managing sets of objects over the full set of needs for distribution, local customization and lifecycle management.